### ISSUE TITLE
AiModelBuilder.BuildAsync fails on CNN/multi-dim-input models · feature index exceeds input dimension

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1468]
## Summary

`AiModelBuilder<T, Tensor<T>, Tensor<T>>.BuildAsync()` throws `ArgumentOutOfRangeException` when used with any prebuilt NN whose first layer has a non-flat input shape (CNN, GRU, Autoencoder, Vision Transformer, etc.). The optimizer's feature-selection pass selects feature indices from the flattened input size but the neural network model's `SetActiveFeatureIndices` validates against `Layers[0].GetInputShape()[0]`, causing a mismatch when the input has more than one spatial/depth dimension.

Found while building the AiDotNet Facade Walkthrough for an instructor-led pair-programming session — the bug breaks the canonical facade chain pattern for any non-tabular use case.

## Reproduction

```csharp
// Tensor input shape [n, depth=3, H=8, W=8] — 3-channel 8×8 images
var (trainX, trainY) = GenerateImages(samples: 60, side: 8, channels: 3, classes: 3);

var layers = new List<ILayer<double>>
{
    new ConvolutionalLayer<double>(outputDepth: 8, kernelSize: 3, stride: 1, padding: 1,
        activationFunction: new ReLUActivation<double>()),
    new MaxPoolingLayer<double>(poolSize: 2, stride: 2),
    new FlattenLayer<double>(),
    new DenseLayer<double>(outputSize: 3, activationFunction: new SoftmaxActivation<double>()),
};
var architecture = new NeuralNetworkArchitecture<double>(
    inputType: InputType.ThreeDimensional,
    taskType: NeuralNetworkTaskType.MultiClassClassification,
    complexity: NetworkComplexity.Simple,
    inputHeight: 8, inputWidth: 8, inputDepth: 3,
    outputSize: 3, layers: layers);
var cnn = new ConvolutionalNeuralNetwork<double>(architecture);

var builder = new AiModelBuilder<double, Tensor<double>, Tensor<double>>();
var adam = new AdamOptimizer<double, Tensor<double>, Tensor<double>>(
    null, new AdamOptimizerOptions<double, Tensor<double>, Tensor<double>>
    { InitialLearningRate = 0.01, MaxIterations = 30 });

await builder
    .ConfigureModel(cnn)
    .ConfigureOptimizer(adam)
    .ConfigureLossFunction(new CategoricalCrossEntropyLoss<double>())
    .ConfigureDataLoader(new InMemoryDataLoader<double, Tensor<double>, Tensor<double>>(trainX, trainY))
    .BuildAsync();
// → throws System.ArgumentOutOfRangeException: Feature index 3 exceeds the input dimension 3.
```

## Stack trace

```
Unhandled exception. System.ArgumentOutOfRangeException: Feature index 3 exceeds the input dimension 3. (Parameter 'featureIndices')
   at AiDotNet.NeuralNetworks.NeuralNetworkBase`1.SetActiveFeatureIndices(IEnumerable`1 featureIndices) in src/NeuralNetworks/NeuralNetworkBase.cs:line 7851
   at AiDotNet.Optimizers.OptimizerBase`3.ApplyFeatureSelection(IFullModel`3 model, List`1 selectedFeatures) in src/Optimizers/OptimizerBase.cs:line 371
   at AiDotNet.Optimizers.OptimizerBase`3.PrepareAndEvaluateSolutionCore(IFullModel`3 solution, OptimizationInputData`3 inputData) in src/Optimizers/OptimizerBase.cs:line 540
   at AiDotNet.Optimizers.OptimizerBase`3.PrepareAndEvaluateSolution(IFullModel`3 solution, OptimizationInputData`3 inputData) in src/Optimizers/OptimizerBase.cs:line 504
   at AiDotNet.Optimizers.AdamOptimizer`3.Optimize(OptimizationInputData`3 inputData) in src/Optimizers/AdamOptimizer.cs:line 205
   at AiDotNet.AiModelBuilder`3.BuildSupervisedInternalAsync(TInput x, TOutput y, CancellationToken cancellationToken) in src/AiModelBuilder.cs:line 3843
   at AiDotNet.AiModelBuilder`3.BuildAsync(CancellationToken cancellationToken) in src/AiModelBuilder.cs:line 1685
```

## Root cause analysis

In `OptimizerBase.PrepareAndEvaluateSolutionCore` (line 515-540):

```csharp
int totalFeatures = InputHelper<T, TInput>.GetInputSize(inputData.XTrain);  // → flat size, e.g. 3*8*8 = 192 for CNN input
selectedFeaturesIndices = Enumerable.Range(0, totalFeatures).ToList();      // or RandomlySelectFeatures(...)
ApplyFeatureSelection(solution, selectedFeaturesIndices);
```

`ApplyFeatureSelection` then calls `model.SetActiveFeatureIndices(selectedFeatures)`, and in `NeuralNetworkBase.SetActiveFeatureIndices` (line 7831-7855):

```csharp
int inputDimension = 0;
if (Layers.Count > 0)
    inputDimension = Layers[0].GetInputShape()[0];   // → 3 for CNN (depth), not 192

foreach (var index in featureIndices)
{
    if (inputDimension > 0 && index >= inputDimension)
        throw new ArgumentOutOfRangeException(...);   // boom at index >= 3
}
```

**The mismatch:** the optimizer treats \"features\" as flat-input columns (range 0..flatSize), but the neural network treats \"features\" as the first dimension of the first layer's input shape (depth for CNN, sequence-length for RNN, etc.). For any model with non-1D input, these counts diverge and any index >= layer-input-dim throws.

## Affected models (confirmed)

- `ConvolutionalNeuralNetwork<T>` — input depth > 1 fails immediately; even depth=1 fails for the same reason (totalFeatures = H*W > 1)
- `GRUNeuralNetwork<T>` / `LSTMNeuralNetwork<T>` — totalFeatures = seqLen * featuresPerStep > featuresPerStep
- Anything else with `InputType.TwoDimensional`/`ThreeDimensional`/`FourDimensional` — Vision Transformer, ResNet, DenseNet, EfficientNet, etc.
- `Autoencoder<T>` has a SEPARATE validation issue: requires first layer's `InputShape == last layer's OutputShape`, but `DenseLayer(int outputSize)` ctor lazy-initializes input shape to `[-1]`, so the equality check fails before training.

## Workaround currently in use

For the W7A/W7B training-program demos and the FineTuneLab demo, I bypass `BuildAsync` and call `model.Train(features, labels)` directly in an explicit epoch loop. That works, but it defeats the purpose of the facade pattern — we lose `ConfigureOptimizer`, `ConfigureLossFunction`, etc. flowing through one chain. Today's developer pair-programming session can't demonstrate the facade for any CV/sequence model as a result.

## Proposed fix

The cleanest fix is to make `OptimizerBase.PrepareAndEvaluateSolutionCore` model-aware: ask the model for its expected feature index range rather than computing from the input tensor's flat size. Something like:

```csharp
int totalFeatures = (solution is IFeatureAware fa) ? fa.GetFeatureCount() : InputHelper<T, TInput>.GetInputSize(inputData.XTrain);
```

Or alternatively, in `NeuralNetworkBase.SetActiveFeatureIndices`, treat indices >= `Layers[0].InputShape[0]` as no-ops for layers that don't support feature subsetting (CNN, RNN, etc.) rather than throwing. The semantics of \"select a subset of input features\" don't really apply to spatial/sequence input anyway.

Either approach unblocks the facade pattern for the entire prebuilt-CNN / Vision Transformer / Autoencoder family.

## Version

Locally building against `main` (commit current as of $(date '+%Y-%m-%d')). Confirmed reproducible.

## Severity

High — blocks the facade pattern (the library's documented preferred API per CLAUDE.md) for every CV/sequence/autoencoder prebuilt model. Tabular flat-input models like the generic `NeuralNetwork<T>` work fine.
